From c942817f500410bf111528289d61a0274c0779e8 Mon Sep 17 00:00:00 2001
From: Howard Lin <howard-yh.lin@mediatek.com>
Date: Fri, 15 Sep 2023 19:22:23 +0800
Subject: [PATCH] GENIO: board: mediatek: add AB firmware updates support

This allows the EFI capsule to select which slot would be updated and
booted by firmware updates when the 'MEDIATEK_IOT_AB_BOOT_SUPPORT'
is enabled.

Change-Id: Ia0aadf2782f3ee5d9b0808546674810b0965b7d2
Signed-off-by: Howard Lin <howard-yh.lin@mediatek.com>
---
 arch/arm/lib/bootm.c                   |  3 +++
 board/mediatek/mt8188/mt8390_evk.c     | 25 +++++++++++++++++++++
 board/mediatek/mt8195/mt8195_demo.c    | 30 ++++++++++++++++++++++++++
 board/mediatek/mt8365/mt8365_pumpkin.c | 25 +++++++++++++++++++++
 board/mediatek/mt8370/mt8370_evk.c     | 25 +++++++++++++++++++++
 cmd/Makefile                           |  2 ++
 cmd/iot_ab_cmd.c                       | 26 ++++++++++++++++++++++
 common/main.c                          |  3 +++
 lib/efi_loader/efi_capsule.c           | 17 ++++++++++++++-
 9 files changed, 155 insertions(+), 1 deletion(-)
 create mode 100644 cmd/iot_ab_cmd.c

diff --git a/arch/arm/lib/bootm.c b/arch/arm/lib/bootm.c
index 9f086f3b90..71c730415e 100644
--- a/arch/arm/lib/bootm.c
+++ b/arch/arm/lib/bootm.c
@@ -302,6 +302,9 @@ static void boot_jump_linux(bootm_headers_t *images, int flag)
 		(ulong) kernel_entry);
 	bootstage_mark(BOOTSTAGE_ID_RUN_OS);
 
+	if (IS_ENABLED(CONFIG_MEDIATEK_IOT_AB_BOOT_SUPPORT))
+		iot_ab_boot_complete();
+
 	announce_and_cleanup(fake);
 
 	if (!fake) {
diff --git a/board/mediatek/mt8188/mt8390_evk.c b/board/mediatek/mt8188/mt8390_evk.c
index b007563d5e..947e6cd14e 100644
--- a/board/mediatek/mt8188/mt8390_evk.c
+++ b/board/mediatek/mt8188/mt8390_evk.c
@@ -11,6 +11,7 @@
 #include <asm/io.h>
 #include <linux/kernel.h>
 #include <linux/arm-smccc.h>
+#include <iot_ab.h>
 
 #define MT8390_UPDATABLE_IMAGES	5
 
@@ -18,8 +19,13 @@
 static struct efi_fw_image fw_images[MT8390_UPDATABLE_IMAGES] = {0};
 
 struct efi_capsule_update_info update_info = {
+#if defined(CONFIG_MEDIATEK_IOT_AB_BOOT_SUPPORT)
+	.dfu_string = "mmc 0=bl2.img raw 0x0 0x2000 mmcpart 1;"
+			"fip.bin part 0 %d;firmware.vfat part 0 %d;u-boot-env.bin raw 0x0 0x2000 mmcpart 2",
+#else
 	.dfu_string = "mmc 0=bl2.img raw 0x0 0x2000 mmcpart 1;"
 			"fip.bin part 0 1;firmware.vfat part 0 3;u-boot-env.bin raw 0x0 0x2000 mmcpart 2",
+#endif
 	.images = fw_images,
 };
 
@@ -69,6 +75,25 @@ void mediatek_capsule_update_board_setup(void)
 		fw_images[4].fw_name = u"GENIO-700-EVK-ENV";
 	}
 }
+
+#if defined(CONFIG_MEDIATEK_IOT_AB_BOOT_SUPPORT)
+void set_dfu_alt_info(char *interface, char *devstr)
+{
+	char alt[BOOTCTRL_DFU_ALT_LEN] = {0};
+	const char *s = env_get(BOOTCTRL_ENV);
+
+	if (s) {
+		if (!strcmp(s, "a")) {
+			if (sprintf(alt, update_info.dfu_string, BOOTCTRL_FIP_NUM + PART_BOOT_B, BOOTCTRL_FW_NUM + PART_BOOT_B) < 0)
+				return;
+		} else if(!strcmp(s, "b")) {
+			if (sprintf(alt, update_info.dfu_string, BOOTCTRL_FIP_NUM, BOOTCTRL_FW_NUM) < 0)
+				return;
+		}
+		env_set("dfu_alt_info", alt);
+	}
+}
+#endif
 #endif /* CONFIG_EFI_HAVE_CAPSULE_SUPPORT && CONFIG_EFI_PARTITION */
 
 int board_init(void)
diff --git a/board/mediatek/mt8195/mt8195_demo.c b/board/mediatek/mt8195/mt8195_demo.c
index 9edddd3088..3f0e12db80 100644
--- a/board/mediatek/mt8195/mt8195_demo.c
+++ b/board/mediatek/mt8195/mt8195_demo.c
@@ -11,6 +11,7 @@
 #include <asm/io.h>
 #include <linux/kernel.h>
 #include <linux/arm-smccc.h>
+#include <iot_ab.h>
 
 #define MT8195_UPDATABLE_IMAGES	5
 
@@ -18,12 +19,22 @@
 static struct efi_fw_image fw_images[MT8195_UPDATABLE_IMAGES] = {0};
 
 struct efi_capsule_update_info update_info = {
+#if defined(CONFIG_MEDIATEK_IOT_AB_BOOT_SUPPORT)
+#if defined(CONFIG_UFS_MEDIATEK)
+	.dfu_string = "ufs 0=bl2.img raw 0x0 0x400 dev 0;"
+			"fip.bin part 2 %d;firmware.vfat part 2 %d;u-boot-env.bin raw 0x0 0x400 dev 1",
+#else
+	.dfu_string = "mmc 0=bl2.img raw 0x0 0x2000 mmcpart 1;"
+			"fip.bin part 0 %d;firmware.vfat part 0 %d;u-boot-env.bin raw 0x0 0x2000 mmcpart 2",
+#endif
+#else
 #if defined(CONFIG_UFS_MEDIATEK)
 	.dfu_string = "ufs 0=bl2.img raw 0x0 0x400 dev 0;"
 			"fip.bin part 2 1;firmware.vfat part 2 3;u-boot-env.bin raw 0x0 0x400 dev 1",
 #else
 	.dfu_string = "mmc 0=bl2.img raw 0x0 0x2000 mmcpart 1;"
 			"fip.bin part 0 1;firmware.vfat part 0 3;u-boot-env.bin raw 0x0 0x2000 mmcpart 2",
+#endif
 #endif
 	.images = fw_images,
 };
@@ -122,6 +133,25 @@ void mediatek_capsule_update_board_setup(void)
 		fw_images[4].fw_name = u"GENIO-1200-EVK-UFS-ENV";
 	}
 }
+
+#if defined(CONFIG_MEDIATEK_IOT_AB_BOOT_SUPPORT)
+void set_dfu_alt_info(char *interface, char *devstr)
+{
+	char alt[BOOTCTRL_DFU_ALT_LEN] = {0};
+	const char *s = env_get(BOOTCTRL_ENV);
+
+	if (s) {
+		if (!strcmp(s, "a")) {
+			if (sprintf(alt, update_info.dfu_string, BOOTCTRL_FIP_NUM + PART_BOOT_B, BOOTCTRL_FW_NUM + PART_BOOT_B) < 0)
+				return;
+		} else if(!strcmp(s, "b")) {
+			if (sprintf(alt, update_info.dfu_string, BOOTCTRL_FIP_NUM, BOOTCTRL_FW_NUM) < 0)
+				return;
+		}
+		env_set("dfu_alt_info", alt);
+	}
+}
+#endif
 #endif /* CONFIG_EFI_HAVE_CAPSULE_SUPPORT && CONFIG_EFI_PARTITION */
 
 int board_init(void)
diff --git a/board/mediatek/mt8365/mt8365_pumpkin.c b/board/mediatek/mt8365/mt8365_pumpkin.c
index 3804a14cfb..f094d56a5f 100644
--- a/board/mediatek/mt8365/mt8365_pumpkin.c
+++ b/board/mediatek/mt8365/mt8365_pumpkin.c
@@ -11,6 +11,7 @@
 #include <asm/io.h>
 #include <linux/kernel.h>
 #include <linux/arm-smccc.h>
+#include <iot_ab.h>
 
 #define MT8365_UPDATABLE_IMAGES	5
 
@@ -18,8 +19,13 @@
 static struct efi_fw_image fw_images[MT8365_UPDATABLE_IMAGES] = {0};
 
 struct efi_capsule_update_info update_info = {
+#if defined(CONFIG_MEDIATEK_IOT_AB_BOOT_SUPPORT)
+	.dfu_string = "mmc 0=bl2.img raw 0x0 0x2000 mmcpart 1;"
+			"fip.bin part 0 %d;firmware.vfat part 0 %d;u-boot-env.bin raw 0x0 0x2000 mmcpart 2",
+#else
 	.dfu_string = "mmc 0=bl2.img raw 0x0 0x2000 mmcpart 1;"
 			"fip.bin part 0 1;firmware.vfat part 0 3;u-boot-env.bin raw 0x0 0x2000 mmcpart 2",
+#endif
 	.images = fw_images,
 };
 
@@ -69,6 +75,25 @@ void mediatek_capsule_update_board_setup(void)
 		fw_images[4].fw_name = u"GENIO-350-EVK-ENV";
 	}
 }
+
+#if defined(CONFIG_MEDIATEK_IOT_AB_BOOT_SUPPORT)
+void set_dfu_alt_info(char *interface, char *devstr)
+{
+	char alt[BOOTCTRL_DFU_ALT_LEN] = {0};
+	const char *s = env_get(BOOTCTRL_ENV);
+
+	if (s) {
+		if (!strcmp(s, "a")) {
+			if (sprintf(alt, update_info.dfu_string, BOOTCTRL_FIP_NUM + PART_BOOT_B, BOOTCTRL_FW_NUM + PART_BOOT_B) < 0)
+				return;
+		} else if(!strcmp(s, "b")) {
+			if (sprintf(alt, update_info.dfu_string, BOOTCTRL_FIP_NUM, BOOTCTRL_FW_NUM) < 0)
+				return;
+		}
+		env_set("dfu_alt_info", alt);
+	}
+}
+#endif
 #endif /* CONFIG_EFI_HAVE_CAPSULE_SUPPORT && CONFIG_EFI_PARTITION */
 
 int board_init(void)
diff --git a/board/mediatek/mt8370/mt8370_evk.c b/board/mediatek/mt8370/mt8370_evk.c
index d3fdc42aa0..00e943b67c 100644
--- a/board/mediatek/mt8370/mt8370_evk.c
+++ b/board/mediatek/mt8370/mt8370_evk.c
@@ -11,6 +11,7 @@
 #include <asm/io.h>
 #include <linux/kernel.h>
 #include <linux/arm-smccc.h>
+#include <iot_ab.h>
 
 #define MT8370_UPDATABLE_IMAGES	5
 
@@ -18,8 +19,13 @@
 static struct efi_fw_image fw_images[MT8370_UPDATABLE_IMAGES] = {0};
 
 struct efi_capsule_update_info update_info = {
+#if defined(CONFIG_MEDIATEK_IOT_AB_BOOT_SUPPORT)
+	.dfu_string = "mmc 0=bl2.img raw 0x0 0x2000 mmcpart 1;"
+			"fip.bin part 0 %d;firmware.vfat part 0 %d;u-boot-env.bin raw 0x0 0x2000 mmcpart 2",
+#else
 	.dfu_string = "mmc 0=bl2.img raw 0x0 0x2000 mmcpart 1;"
 			"fip.bin part 0 1;firmware.vfat part 0 3;u-boot-env.bin raw 0x0 0x2000 mmcpart 2",
+#endif
 	.images = fw_images,
 };
 
@@ -69,6 +75,25 @@ void mediatek_capsule_update_board_setup(void)
 		fw_images[4].fw_name = u"GENIO-510-EVK-ENV";
 	}
 }
+
+#if defined(CONFIG_MEDIATEK_IOT_AB_BOOT_SUPPORT)
+void set_dfu_alt_info(char *interface, char *devstr)
+{
+	char alt[BOOTCTRL_DFU_ALT_LEN] = {0};
+	const char *s = env_get(BOOTCTRL_ENV);
+
+	if (s) {
+		if (!strcmp(s, "a")) {
+			if (sprintf(alt, update_info.dfu_string, BOOTCTRL_FIP_NUM + PART_BOOT_B, BOOTCTRL_FW_NUM + PART_BOOT_B) < 0)
+				return;
+		} else if(!strcmp(s, "b")) {
+			if (sprintf(alt, update_info.dfu_string, BOOTCTRL_FIP_NUM, BOOTCTRL_FW_NUM) < 0)
+				return;
+		}
+		env_set("dfu_alt_info", alt);
+	}
+}
+#endif
 #endif /* CONFIG_EFI_HAVE_CAPSULE_SUPPORT && CONFIG_EFI_PARTITION */
 
 int board_init(void)
diff --git a/cmd/Makefile b/cmd/Makefile
index 6e87522b62..c06ff48fa1 100644
--- a/cmd/Makefile
+++ b/cmd/Makefile
@@ -194,6 +194,8 @@ obj-$(CONFIG_CMD_ETHSW) += ethsw.o
 obj-$(CONFIG_CMD_AXI) += axi.o
 obj-$(CONFIG_CMD_PVBLOCK) += pvblock.o
 
+obj-$(CONFIG_MEDIATEK_IOT_AB_BOOT_SUPPORT) += iot_ab_cmd.o
+
 # Power
 obj-$(CONFIG_CMD_PMIC) += pmic.o
 obj-$(CONFIG_CMD_REGULATOR) += regulator.o
diff --git a/cmd/iot_ab_cmd.c b/cmd/iot_ab_cmd.c
new file mode 100644
index 0000000000..ccbe23ed21
--- /dev/null
+++ b/cmd/iot_ab_cmd.c
@@ -0,0 +1,26 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+/*
+ * Copyright (C) 2023 MediaTek Inc.
+ * Author: Howard Lin <howard-yh.lin@mediatek.com>
+ */
+
+#include <common.h>
+#include <command.h>
+
+static int do_iot_ab_cmd(struct cmd_tbl *cmdtp, int flag, int argc,
+			char *const argv[])
+{
+	if (argc != 1)
+		return CMD_RET_USAGE;
+
+	iot_ab_boot_select();
+
+	printf("The slot is changed successfully!\n");
+	return CMD_RET_SUCCESS;
+}
+
+U_BOOT_CMD(iot_ab, 1, 0, do_iot_ab_cmd,
+	   "Change boot slot to A or B for MediaTek IoT Yocto.",
+	   "\n"
+	   "    - Change boot slot of boot_ctrl data from metadata on the device.\n"
+);
diff --git a/common/main.c b/common/main.c
index 682f3359ea..baa2ddb7b4 100644
--- a/common/main.c
+++ b/common/main.c
@@ -60,6 +60,9 @@ void main_loop(void)
 			efi_launch_capsules();
 	}
 
+	if (IS_ENABLED(CONFIG_MEDIATEK_IOT_AB_BOOT_SUPPORT))
+		iot_ab_boot_slot(0);
+
 	s = bootdelay_process();
 	if (cli_process_fdt(&s))
 		cli_secure_boot_cmd(s);
diff --git a/lib/efi_loader/efi_capsule.c b/lib/efi_loader/efi_capsule.c
index a6b98f066a..10da0443b1 100644
--- a/lib/efi_loader/efi_capsule.c
+++ b/lib/efi_loader/efi_capsule.c
@@ -1104,6 +1104,9 @@ efi_status_t efi_launch_capsules(void)
 	u16 **files;
 	unsigned int nfiles, index, i;
 	efi_status_t ret;
+#if defined(CONFIG_MEDIATEK_IOT_AB_BOOT_SUPPORT)
+	unsigned int boot_ab = 0;
+#endif
 
 	if (check_run_capsules() != EFI_SUCCESS)
 		return EFI_SUCCESS;
@@ -1134,10 +1137,17 @@ efi_status_t efi_launch_capsules(void)
 			if (ret != EFI_SUCCESS)
 				log_err("Applying capsule %ls failed.\n",
 					files[i]);
+#if !defined(CONFIG_MEDIATEK_IOT_AB_BOOT_SUPPORT)
 			else
 				log_info("Applying capsule %ls succeeded.\n",
 					 files[i]);
-
+#else
+			else {
+				log_info("Applying capsule %ls succeeded.\n",
+					 files[i]);
+				boot_ab++;
+			}
+#endif
 			/* create CapsuleXXXX */
 			set_capsule_result(index, capsule, ret);
 
@@ -1153,6 +1163,11 @@ efi_status_t efi_launch_capsules(void)
 	}
 	efi_capsule_scan_done();
 
+	if (IS_ENABLED(CONFIG_MEDIATEK_IOT_AB_BOOT_SUPPORT)) {
+		if (boot_ab == nfiles)
+			iot_ab_boot_select();
+	}
+
 	for (i = 0; i < nfiles; i++)
 		free(files[i]);
 	free(files);
-- 
2.18.0

